### ISSUE TITLE
Determine pkg-config develop requirements for packages

### ISSUE THREAD

**Issue by Begasus** [haikuports/haikuports#11268]
Following up on a PR for curl where in first place it seemed to be missing required _devel packages in REQUIRES_devel: https://github.com/haikuports/haikuports/pull/11262 where @korli fixed this in: https://github.com/haikuports/haikuports/commit/de1a699f7b33176b9240ad7390248aa36fbb46d8

I'm hitting this (and probably this is the same for several other packages) with building kiwix which requires libzim.

Now libzim (currently) doesn't require the _devel packages for liblzma and libzstd and therefore kiwix will fail detecting libzim:

````
Determining dependency 'libzim' with pkg-config executable '/bin/pkg-config'
env[PKG_CONFIG_PATH]: 
-----------
Called: `/bin/pkg-config --modversion libzim` -> 0
stdout:
9.2.3
-----------
env[PKG_CONFIG_PATH]: 
-----------
Called: `/bin/pkg-config --cflags libzim` -> 1
stderr:
Package libzstd was not found in the pkg-config search path.
Perhaps you should add the directory containing `libzstd.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libzstd', required by 'libzim', not found
-----------
pkg-config error with 'libzim': Could not generate cargs for libzim:
Package libzstd was not found in the pkg-config search path.
Perhaps you should add the directory containing `libzstd.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libzstd', required by 'libzim', not found

Dependency lookup for libzim with method 'pkgconfig' failed: Could not generate cargs for libzim:
Package libzstd was not found in the pkg-config search path.
Perhaps you should add the directory containing `libzstd.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libzstd', required by 'libzim', not found

CMake binary for 1 is not cached
CMake binary missing from cross or native file, or env var undefined.
Trying a default CMake fallback at cmake
Found CMake: /bin/cmake (3.30.5)
Extracting basic cmake information
CMake Toolchain: Calling CMake once to generate the compiler state
Calling CMake (['/bin/cmake']) in /sources/libkiwix-13.1.0/build/meson-private/__CMake_compiler_info__ with:
  - "--trace-expand"
  - "--trace-format=json-v1"
  - "--no-warn-unused-cli"
  - "--trace-redirect=cmake_trace.txt"
  - "-G"
  - "Ninja"
  - "-DCMAKE_TOOLCHAIN_FILE=/sources/libkiwix-13.1.0/build/meson-private/__CMake_compiler_info__/CMakeMesonTempToolchainFile.cmake"
  - "."
Try CMake generator: auto
Calling CMake (['/bin/cmake']) in /sources/libkiwix-13.1.0/build/meson-private/cmake_libzim with:
  - "--trace-expand"
  - "--trace-format=json-v1"
  - "--no-warn-unused-cli"
  - "--trace-redirect=cmake_trace.txt"
  - "-DCMAKE_TOOLCHAIN_FILE=/sources/libkiwix-13.1.0/build/meson-private/cmake_libzim/CMakeMesonToolchainFile.cmake"
  - "."
  -- Module search paths:    ['/boot/system', '/boot/system/develop']
  -- CMake root:             /boot/system/data/cmake
  -- CMake architectures:    []
  -- CMake lib search paths: ['lib', 'lib32', 'lib64', 'libx32', 'share']
Preliminary CMake check failed. Aborting.
Run-time dependency libzim found: NO (tried cmake)

meson.build:38:13: ERROR: Dependency lookup for libzim with method 'pkgconfig' failed: Could not generate cargs for libzim:
Package libzstd was not found in the pkg-config search path.
Perhaps you should add the directory containing `libzstd.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libzstd', required by 'libzim', not found
````

Easy fix, add libzma and libzstd to the REQUIRES_devel from libzim.

Question would be, is it really needed to have them in there? Or do I use the same as is done for curl, and how do I know they are really required for libzim?

**Comment by Begasus** [comment]
Related issue upstream at meson? https://github.com/mesonbuild/meson/issues/3970

**Comment by mtl1979** [comment]
The way I check for true dependencies for shared libraries is try to link against the shared library and run the resulting binary... This is the only way to see if the indirect dependencies are actually needed. In Haiku and few other operating systems shared libraries are allowed to have undefined symbols, thus merely linking doesn't cause an error.

Whenever possible, I try to avoid using tools like pkg-config, because in modern build scripts, those are almost always redundant and there is already better way of getting the same information.

**Comment by Begasus** [comment]
Most sources use pkg-config or cmake to find required libraries/headers, I don't think you can by-pass that (or nuke the check in the sources, but that shouldn't be the way to go).

**Comment by mtl1979** [comment]
> Most sources use pkg-config or cmake to find required libraries/headers, I don't think you can by-pass that (or nuke the check in the sources, but that shouldn't be the way to go).

Some sources use both pkg-config and cmake and that is redundant... pkg-config is usually only needed when using hand-written configure files or only makefiles.

I have only used meson once and it failed with even finding common system libraries, so maybe it is not the smartest build system and needs some help...

**Comment by korli** [comment]
The question is more whether the library headers depend on third-party headers. If it's the case these libs need to be added in the devel requirements.

**Comment by mtl1979** [comment]
Because haikuporter builds recipe in chroot, the build should fail if headers depend on other headers not available inside the chroot environment... Linking against unnecessary libraries however doesn't throw any errors.

**Comment by Begasus** [comment]
> The question is more whether the library headers depend on third-party headers. If it's the case these libs need to be added in the devel requirements.

So in the case for libzim, I should grep in the source if there is any reference for `#include <zstd.h>` as an example? For the record, nothing pops up when searching in those headers for libzim.

**Comment by Begasus** [comment]
> Because haikuporter builds recipe in chroot, the build should fail if headers depend on other headers not available inside the chroot environment... Linking against unnecessary libraries however doesn't throw any errors.

This is already the case.

**Comment by Begasus** [comment]
> not the source, only the headers contained in the devel package.

Yeah, expressed myself wrong there, checked the directory "zim" in /system/develop/headers for that, so if there is no trace of the header it shouldn't be included in the pkgconfig file for libzim's section Requires.private?

**Comment by Begasus** [comment]
Thanks! I'm sure this will clear up some issues when users building with haikuporter and run into this. 👍

**Comment by korli** [comment]
> Thanks! I'm sure this will clear up some issues when users building with haikuporter and run into this. 👍

It might be a good idea to have a shell scriptlet for this.

**Comment by Begasus** [comment]
> It might be a good idea to have a shell scriptlet for this.

Way over my skills (I think), and too much side-tracks :) but yes, some automation would be handy here, I've been told to check with pkg-config on required packages years ago (can't remember exactly who it was), but following that blindly isn't the correct way to go it seems with this clarification.

**Comment by mtl1979** [comment]
`hgrep` can find text in header files, but you need to know what to pass to it...

**Comment by Begasus** [comment]
Already got a alias for hgrepall (@OscarL  thanks for that!) that could do that trick. :)

**Comment by mtl1979** [comment]
> Already got a alias for hgrepall (@OscarL thanks for that!) that could do that trick. :)

I remember making loud noise about keeping lgrep and hgrep around 16 years ago... Quite useful shell scripts/macros...
